File: Data Structures & Algorithms/spiral-matrix/submission-1.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int l = 0, d = n, t = 0, r = matrix[0].size();
        vector<int> ans;
        while(l<r && t < d){
            for(int i = l;i<r;i++){
                ans.push_back(matrix[t][i]);
            }
            t ++;
            for(int i = t;i<d;i++){
                ans.push_back(matrix[i][r-1]);
            }
            r--;
            if (!(l < r && t < d)) {
                break;
            }
            for(int i = r-1;i>=l;i--){
                ans.push_back(matrix[d-1][i]);
            }
            d--;
            for(int i = d-1;i>=t;i--){
                ans.push_back(matrix[i][l]);
            }
            l++;
        }
        return ans;
    }
};

File: Data Structures & Algorithms/minimum-interval-including-query/submission-1.java
class Solution {
    public int[] minInterval(int[][] intervals, int[] queries) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        
        PriorityQueue<int[]> q = new PriorityQueue<>((a, b) -> {
            if (a[0] == b[0]) {
                return Integer.compare(a[1], b[1]);
            }
            return Integer.compare(a[0], b[0]);
        });

        Map<Integer, Integer> res = new HashMap<>();
        int i = 0;
        int intervalsLen = intervals.length;

        for (int que : Arrays.stream(queries).sorted().toArray()) {
            
            while (i < intervalsLen && intervals[i][0] <= que) {
                int l = intervals[i][0];
                int r = intervals[i][1];
                q.offer(new int[]{r - l + 1, r});
                i++;
            }

            while (!q.isEmpty() && q.peek()[1] < que) {
                q.poll();
            }

            res.put(que, q.isEmpty() ? -1 : q.peek()[0]);
        }

        int[] ans = new int[queries.length];
        for (int j = 0; j < queries.length; j++) {
            ans[j] = res.get(queries[j]);
        }

        return ans;
    }
}